### ISSUE TITLE
maintenance aside, what should the future of govendor be?

### ISSUE THREAD

**Issue by kardianos** [kardianos/govendor#292]
`govendor` started April of 2015.

The govendor tool started as an attempt to unify the vendor.json file or make a single tool that does what people need it to do. Apparently it failed at that. I would like a standard tool to succeed. I just don't want another npm. I'm sceptical of any attempt to update to something other then latest major version. I'm skeptical of version solvers in my software tool chain. It could be I'm wrong on these counts.

Right now govendor solves my personal problems, and the problems others have as well. In light of golang/dep, I'm trying to decide the fate of this tool.

 * Should it be sunsetted in favor of golang/dep?
 * Should I work on a version 2 that builds on what is present, potentially simplifies the commands and CLI interface, is more reliable with fetches?
 * Should I just fix the rough edges and inconsistencies that are found in govendor and continue to keep it alive?

If golang/dep becomes official and integrated into the tool chain it may be best to simply sunset govendor and put engineering effort into that.

I have serious concern that dep will be too much like npm or pip. Already it has transitioned from a well defined design document into gradually becoming something Sam has always advocated for:

 * A separate lock and manifest file.
 * The manifest file is now manually edited.
 * The vendor folder may go away.
 * Dependencies automatically resolve, pulling many versions from the VCS often causing servers such as github to reject requests. I don't think the solution is to require people to login to a github account to enable resolving dependencies.
 * Dependencies automatically resolve. Ensuring you are updating to the latest version may mean specifying master in the revision or manually checking each dependency for the latest tag. This isn't the end of the world, but is a grave concern to me. I don't care if something is incompatible. If it isn't the latest it isn't maintained. I fear unknowingly staying on an old version of a dep even after I thought I updated.
 * I don't yet feel like there is complete by-in from Sam as the `gps` package isn't yet in the same repository, despite saying repeatedly that velocity is paramount. Putting them together would have been one of the first things I would have done for speed alone. He is concerned that doing so will "kill" the gps package. This doesn't feel like buy-in to me. (There is a plan to now integrate gps into dep, but Sam's initial response continues to concern me.)

I'm not aware of an up-to-date design document for golang/dep anymore. This again concerns me.

Any comments welcome.

---

EDIT: Russ wrote a reply to the dep roadmap message here: https://groups.google.com/d/msg/golang-nuts/PaGu2s9knao/Bq4vmFh7AgAJ which clarifies the go position more. Sam's response to that email continues to concern me. I'm currently leaning heavily to continue developing govendor for the time being.

**Comment by myitcv** [comment]
@kardianos no specific comments, but I will follow this thread as an interested bystander. 

I've not yet had the chance to evaluate the design docs or the `golang/dep` tool itself; I'm still using a `GOPATH` augmentation approach with some simple backing scripts that maintain a pin file. Helpful therefore to see your evaluation here at least in a context, `govendor` that I do currently understand.

**Comment by adg** [comment]
I am a bit nonplussed by this. I am having trouble connecting the statements here with my understanding of things, and also connecting "dep will be too much like npm or pip"  "[dep is] becoming something Sam has always advocated for" with the other statements in this issue.

> The manifest file is now manually edited.

This was not something that Sam advocated for. In fact, he was against it. I took some convincing myself, but I think the end result is something that will work well in practice.

> pulling many versions from the VCS often causing servers such as github to reject requests.

For end users this should be no different a workload than what 'go get' does. For maintainers this should only happen when updating or upgrading dependencies.

> The vendor folder may go away.

It's true that the vendor folder may go away at some point, but that would be as a result of reworking the go tool as a whole, an effort that is being driven by @rsc. Dep will continue to use vendor for the foreseeable future, as that is what is afforded us by the go tool. Project maintainers will continue to want to check their dependencies into their repositories, and that's a workflow we must continue to support.

> Dependencies automatically resolve. Ensuring you are updating to the latest version may mean specifying master in the revision or manually checking each dependency for the latest tag.

I don't really understand this. The ecosystem we're heading toward is one where the latest released version is the preferred one. Is that not how you see it?

> I don't yet feel like there is complete by-in from Sam as the gps package isn't yet in the same repository

Again, this doesn't resonate with me. We have spoken a few times about rewriting the parts of gps that dep uses and Sam doesn't seem particularly attached to it. As you say, gps is now being moved into the dep repo to increase development velocity.

> I'm not aware of an up-to-date design document for golang/dep anymore.

This is a problem and I'll try to address it personally.

As for govendor, I'd suggest you continue to maintain it to support your users. Hopefully dep will not be a failure and we can all move over to it.

**Comment by kardianos** [comment]
Thanks @adg for the comments and your time. I think I conflated different issues, and I'm probably wrong on others. I think I was describing three different things:

 1. golang/dep doesn't have a design document currently, and the design has been changing without updating one. In makes it difficult to assess for sure that golang/dep will be able to fully replace govendor. This is probably the only point that pertains to this issue (what features to consider / what bugs to fix). I think this took me the lengths of the document to come to, and I failed to go back and revise my earlier text.
 2. I have a personal concern, that should not involve govendor (this was my mistake), that Sam shaping golang/dep in his image and not a decided design.  I'm just seeing many of the things he originally advocated for a year ago (including manually editing the manifest file) slowly seep back into golang/dep. It isn't necessarily bad, maybe he was right, but that coupled with a lack of living design document feels.. fishy. I personally have this concern. It is very likely I'm wrong. Either way this doesn't really belong here, I'm sorry.
 3. I have a technical concern that resolving dependencies directly from a VCS the way dep currently does it without a central metadata store is a bad design. It feels like we should either focus on reducing the solver complexity or admit we need a central metadata server. I did not state this concern well at all, and again, this was my mistake. And again, doesn't really belong in this venue/issue.

I don't want dep to fail. I do want all go users to unify around a single tool, hopefully dep (or a successor in the go cmd).

Going forward in this issue, I want to focus on:

 1. What is the proposed design of golang/dep? Will it satisfy current govendor users when it is complete?
 2. What are the scope of changes that I should consider going into govendor in light of a possible move to golang/dep in the future?

Unless someone else raises other comments, I'll likely close this issue and open up two corresponding issues pointed directly at these points.
